tests/opentelemetry-docker-tests/tests/prometheus-remote-write-cortex/prometheus_remote_write_cortex_test.py

```diff
@@ -0,0 +1,71 @@
+from opentelemetry import metrics
+from opentelemetry.exporter.prometheus_remote_write import (
+    PrometheusRemoteWriteMetricsExporter,
+)
+from opentelemetry.test.test_base import TestBase
+
+
+def observer_callback(observer):
+    array = [1.0, 15.0, 25.0, 26.0]
+    for (index, usage) in enumerate(array):
+        labels = {"test_label": str(index)}
+        observer.observe(usage, labels)
+
+
+class TestPrometheusRemoteWriteExporterCortex(TestBase):
+    def setUp(self):
+        super().setUp
+        self.exporter = PrometheusRemoteWriteMetricsExporter(
+            endpoint="http://localhost:9009/api/prom/push",
+            headers={"X-Scope-Org-ID": "5"},
+        )
+        self.labels = {"environment": "testing"}
+        self.meter = self.meter_provider.get_meter(__name__)
+        metrics.get_meter_provider().start_pipeline(
+            self.meter, self.exporter, 1,
+        )
+
+    def test_export_counter(self):
```

Comment:
What is being tested here? There is no assertion. If what you want to test is that no exceptions were raised use [`fail`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.fail). More information [here](https://stackoverflow.com/a/4319870).

Reply:
We originally had the pattern you linked but were recommended internally to let the test Error out naturally since that fails as well. The main issue we found with catch ExceptionType was it only handled one Exception type. Is this wrong?

Reply:
Yes, it is wrong.

Every test case *must* have at least one assertion, since that indicates what is the intention of the test case. Even if your intention is _test that this code does not raise an exception_ then that must be made obvious by using an assertion. It is not the same thing when a test case fails because, let's say, an `IndexError` was raised because a queried sequence was too short than when a test case fails with an `AssertionError` because some counter was less than expected. The first situation tells you that the actual testing was not completed because something else went wrong in the process of testing. The second one tells you that you were able to reach the point where you could do the testing that you wanted.

For example, imagine you are testing the performance of a server. As every test case, it begins with a criteria that must be met for the test to pass. Let's say, the server must be able to reply back to 100 requests in a second. So, your test case basically consists in this:

1. Start the server
2. Send the server 100 requests in one second
3. *Assert* that you got 100 replies

If your test case fails with an `AssertionError` whose message is something like `We only got 94 replies`, then you know that the server is not performing as expected **and your testing process can be considered complete**. If your test case fails with `ImportError` then you can possibly figure out that some dependency of the server code is not installed in your testing environment. But the most important thing that you know from this second scenario is that **you did not test what you wanted to test**. That means you can not yet make any conclusions about the performance of the server because you got an exception different from an `AssertionError`. Only this kind of exception should be telling you that the test case failed because what you wanted to test did not comply with what was expected from it. A testing process is only complete if there are 0 failures or if every failure that there is is caused by an `AssertionError`. Every other exception raised is telling you, we were not able to perform the test, fix the issue and **run again**.

This makes sense because a tester can't report back to a developer any other result than a pass or a failure with an `AssertionError`. If a tester reports back an `ImportError` the developer will argue rightly that the code under test was not tested so it can't be considered as being at fault.

So, to summarize, it is perfectly fine that your testing criteria is _this code must run without raising any exception_. Even so, you must **assert** that, so that it is obvious to anyone that reads your test case what the intention of the test case is.

Something else, Python allows you to [catch multiple exceptions types](https://docs.python.org/3/tutorial/errors.html#handling-exceptions). Also keep in mind that what except does is pretty much the same as `isinstance` does, so you can catch multiple exceptions if they share a parent:

``` python
class Parent(Exception):
    pass


class Child0(Parent):
    pass


class Child1(Parent):
    pass


try:
    raise Child0()
except Parent:
    print("Caught a Child0")

try:
    raise Child1()
except Parent:
    print("Caught a Child1")
```
```
Caught a Child0
Caught a Child1
```

Reply:
By the way, pytest makes it confusing because every exception raised in a test case is reported as a failure. In my opinion, this situation should be reported as an `ERROR` instead of a `FAILURE`. An `ERROR` is something that went wrong while getting ready to test something. This is what happens when an exception is raised in a [fixture](https://docs.pytest.org/en/stable/fixture.html?highlight=fixture#pytest-fixtures-explicit-modular-scalable), which is exactly that, code that is run to get something ready for testing. Here is an example:

``` python
from pytest import fixture


@fixture
def preparation():
    1 / 0


def test_case(preparation):

    assert 1 > 0
```
```
pytest test_error.py 
===================================================================================== test session starts ======================================================================================
platform linux -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /home/ocelotl/lightstep/metric_language_equivalent_tests/python/varman
plugins: cov-2.10.0
collected 1 item                                                                                                                                                                               

test_error.py E                                                                                                                                                                          [100%]

============================================================================================ ERRORS ============================================================================================
_________________________________________________________________________________ ERROR at setup of test_case __________________________________________________________________________________

    @fixture
    def preparation():
>       1 / 0
E       ZeroDivisionError: division by zero

test_error.py:6: ZeroDivisionError
=================================================================================== short test summary info ====================================================================================
ERROR test_error.py::test_case - ZeroDivisionError: division by zero
======================================================================================= 1 error in 0.14s =======================================================================================
```

As you can see here, the result of the run is `ERROR` and not `FAILURE` because pytest is telling us that it was unable to do the actual testing because something went wrong when getting ready to run the actual test. `ERROR` means the _tester_ must fix something, `FAILURE` means the _developer_ must fix something.

Maybe someday I'll write a pytest plugin that will make every non `AssertionError` exception to be reported as an `ERROR` :stuck_out_tongue:

Reply:
ahh okay, thanks for the detailed answer! I will fix it rn :D

Reply:
Although I am not the one making the face, thanks for the detailed response. It was very insightful.

Reply:
My pleasure, fantastic job on getting all these PRs done @AzfaarQureshi @shovnik :muscle: